### ISSUE TITLE
[STORM-1946] ShellBolt.java - On busy system BoltHeartbeatTimerTask fires before setHeartbeat() is executed

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#5728]
<p>When storm stars a large number of ShellBolt-s that consume a lot of CPU time to initialize, it creates a lot of contention between processes for CPU resource. That leads to <a href="https://github.com/apache/storm/blob/master/storm-core/src/jvm/org/apache/storm/task/ShellBolt.java#L142" class="external-link" target="_blank" rel="nofollow noopener">BoltHeartbeatTimerTask</a> being fired up after 1 second delay <em>before</em> <a href="https://github.com/apache/storm/blob/master/storm-core/src/jvm/org/apache/storm/task/ShellBolt.java#L145" class="external-link" target="_blank" rel="nofollow noopener">setHeartbeat()</a> assigns initial value to <a href="https://github.com/apache/storm/blob/master/storm-core/src/jvm/org/apache/storm/task/ShellBolt.java#L91" class="external-link" target="_blank" rel="nofollow noopener">lastHeartbeatTimestamp</a> variable.<br/>
As a result when <tt>BoltHeartbeatTimeTask</tt> fires up for the first time, <a href="https://github.com/apache/storm/blob/master/storm-core/src/jvm/org/apache/storm/task/ShellBolt.java#L316" class="external-link" target="_blank" rel="nofollow noopener">getLastHeartbeat()</a> returns value of <b>0</b>. This in turn leads bolt to die with <a href="https://github.com/apache/storm/blob/master/storm-core/src/jvm/org/apache/storm/task/ShellBolt.java#L322" class="external-link" target="_blank" rel="nofollow noopener">"subprocess heartbeat timeout"</a> message.<br/>
The fix is to place <tt>setHeartBeat()</tt> <em>before</em> <tt>BoltHeartbeatTimerTask</tt> is created. The patch for this is attached.</p>

---
<details><summary><i>Originally reported by <a title="Slava Andreyev" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slava92">slava92</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-1946" target="_blank">ShellBolt.java - On busy system BoltHeartbeatTimerTask fires before setHeartbeat() is executed</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Slava Andreyev" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slava92">slava92</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2016-07-08T08:24:52+01:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slava92">slava92</a>:</i>
<p>A patch to fix the issue.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user slava92 opened a pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1542" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1542</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-1946" title="ShellBolt.java - On busy system BoltHeartbeatTimerTask fires before setHeartbeat() is executed" class="issue-link" data-issue-key="STORM-1946"><del>STORM-1946</del></a>: initialize lastHeartbeatTimestamp before starting heartbeat task</p>

<p>    When storm stars a large number of ShellBolt-s that consume a lot of CPU time to initialize, it creates a lot of contention between processes for CPU resource. That leads to BoltHeartbeatTimerTask being fired up after 1 second delay before setHeartbeat() assigns initial value to lastHeartbeatTimestamp variable.<br/>
    As a result when BoltHeartbeatTimeTask fires up for the first time, getLastHeartbeat() returns value of 0. This in turn leads bolt to die with "subprocess heartbeat timeout" message.<br/>
    The fix is to place setHeartBeat() before BoltHeartbeatTimerTask is created. The patch for this is attached.</p>


<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/slava92/storm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/slava92/storm</a> master</p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/storm/pull/1542.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1542.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #1542</p>

<hr />
<p>commit d38c40026ab5dac266986896424e6c25a440e512<br/>
Author: Slava Andreyev <sandreyev@ets.org><br/>
Date:   2016-07-06T17:14:09Z</p>

<p>    initialize lastHeartbeatTimestamp before starting heartbeat task</p>

<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1542" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1542</a></p>

<p>    +1 Great finding.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p>Thanks <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slava92" class="user-hover" rel="slava92">Slava Andreyev</a>, I merged into master, 1.x, 1.0.x branches respectively.</p>

**Comment by jira-importer** [comment]
<i>[Epic: <a href="https://github.com/apache/storm/issues?q=in%3Atitle%20STORM-1855%20label%3Aepic">STORM-1855</a>]</i>
